src/cli.c

```diff
@@ -0,0 +1,125 @@
+#include "cli.h"
+#include "logging.h"
+
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <limits.h>
+#include <getopt.h>
+#include <stdio.h>
+#include <errno.h>
+
+#define MAX_HZ (1000000L)
+
+static const char* usage_str = "cretro [-d <0..5>] [-f <FREQUENCY_HZ>] <ROM>";
+
+static int _save_strtol(const char* str_to_conv, int* store_into) {
+    char* end;
+    const long strtol_in = strtol(str_to_conv, &end, 10);
+
+    errno = 0;
+
+    if (end == str_to_conv) {
+        LOG_FATAL("%s: not a decimal number", str_to_conv);
+    } else if ('\0' != *end) {
+        LOG_FATAL("%s: extra characters at end of input: %s", str_to_conv, end);
+    } else if ((LONG_MIN == strtol_in || LONG_MAX == strtol_in) && ERANGE == errno) {
+        LOG_FATAL("%s out of range of type long", str_to_conv);
+    } else if (strtol_in > INT_MAX) {
+        LOG_FATAL("%ld greater than INT_MAX", strtol_in);
+    } else if (strtol_in < INT_MIN) {
+        LOG_FATAL("%ld less than INT_MIN", strtol_in);
+    } else {
+        *store_into = (int)strtol_in;
+        return EXIT_SUCCESS;
+    }
+
+    return EXIT_FAILURE;
+}
+
+static void _handle_arg_frequency(config_t* conf, long frequency) {
+    // calculate delay from hertz input
+    LOG_DEBUG("Frequency [Hz] input specified: %ld.", frequency);
+
+    if (frequency > MAX_HZ) {
+        LOG_ERROR("Provided frequency %ld is higher than allowed frequency %ld. Used Fallback: %ld", frequency, MAX_HZ, MAX_HZ);
+        frequency = MAX_HZ;
+    }
+    else if (frequency <= 0) {
+        LOG_ERROR("Provided frequency %ld is lower or equal to 0. Used Fallback: %ld", frequency, MAX_HZ);
+        frequency = MAX_HZ;
+    }
+
+    long delay = MAX_HZ / frequency;
+    if (delay > INT_MAX)
+        LOG_FATAL("Delay is out of bounds: %d > %d", delay, INT_MAX);
+
+    conf->us_delay = (int) (delay);
+    LOG_INFO("Delay set to %d us.", conf->us_delay);
+}
+
+config_t* cli_config_default(void) {
+    config_t* conf = malloc(sizeof(config_t));
+
+    conf->debug = FATAL;
+    conf->us_delay = 0;
+    conf->rom_path = "";
+
+    return conf;
+}
+
+void cli_config_destroy(config_t *conf) {
+    free(conf);
+}
+
+int cli_config_handle(config_t* const conf, int argc, char **argv) {
+    int c;
+
+    if (argc < 2) {
+        fprintf(stderr, "%s\n", usage_str);
+        return EXIT_FAILURE;
+    }
+
+    // parse all options first
+    int strtol_in;
+    while ((c = getopt(argc, argv, "d:f:")) != -1) {
+        switch (c) {
+            case 'd':
+                if (_save_strtol(optarg, &strtol_in))
+                    return EXIT_FAILURE;
+
+                conf->debug = strtol_in;
+                log_init(conf);
+                break;
+            case 'f':
+                if (_save_strtol(optarg, &strtol_in))
```

Comment:
I think that instead of handling back the `EXIT_FAILURE` from 3 functions to main, we could simply do `exit(EXIT_FAILURE)` at the point where the unrecoverable error happens.

Reply:
But we have to be careful when doing this since, we need to free some resources, e.g., `SDL_Quit`. So we could either have a general function/macro that frees everything up and then does `exit(value)` or we continue with the return chains. I think that both is fine actually. 

Having the general function/macro for "exiting" could also be used to flush std and stderr since they might contain some unflushed things before we exit.

Reply:
I personally like to propagate errors as far back as possible as this makes the control flow a bit more obvious.

Reply:
This is fine in this use case, as the config is parsed immediately when the program starts before SDL allocates anything.
If the CLI fails, the main function fails immediately as well.

Reply:
But the config struct gets `malloc`ed, this has to be freed. Just quitting back to main mitigates the need to free it somewhere else but `main`.

Reply:
My bad :D